source/common/http/conn_manager_utility.cc

```diff
@@ -176,14 +176,49 @@ Network::Address::InstanceConstSharedPtr ConnectionManagerUtility::mutateRequest
     request_headers.insertRequestId().value(uuid);
   }
 
-  if (config.tracingConfig()) {
-    Tracing::HttpTracerUtility::mutateHeaders(request_headers, runtime);
-  }
+  mutateTracingRequestHeader(request_headers, runtime, config);
   mutateXfccRequestHeader(request_headers, connection, config);
 
   return final_remote_address;
 }
 
+void ConnectionManagerUtility::mutateTracingRequestHeader(Http::HeaderMap& request_headers,
+                                                          Runtime::Loader& runtime,
+                                                          ConnectionManagerConfig& config) {
+  if (!config.tracingConfig() || !request_headers.RequestId()) {
+    return;
+  }
+
+  std::string x_request_id = request_headers.RequestId()->value().c_str();
+  uint64_t result;
+  // Skip if x-request-id is corrupted.
+  if (!UuidUtils::uuidModBy(x_request_id, result, 10000)) {
+    return;
+  }
+
+  // Do not apply tracing transformations if we are currently tracing.
+  if (UuidTraceStatus::NoTrace == UuidUtils::isTraceableUuid(x_request_id)) {
+    if (request_headers.ClientTraceId() &&
+        (runtime.snapshot().featureEnabled("tracing.client_enabled",
+                                           config.tracingConfig()->client_sampling_))) {
+      UuidUtils::setTraceableUuid(x_request_id, UuidTraceStatus::Client);
+    } else if (request_headers.EnvoyForceTrace()) {
+      UuidUtils::setTraceableUuid(x_request_id, UuidTraceStatus::Forced);
+    } else if (runtime.snapshot().featureEnabled("tracing.random_sampling",
+                                                 config.tracingConfig()->random_sampling_, result,
+                                                 10000)) {
+      UuidUtils::setTraceableUuid(x_request_id, UuidTraceStatus::Sampled);
+    }
+  }
+
+  if (!runtime.snapshot().featureEnabled("tracing.global_enabled",
```

Comment:
Although not related to this PR, this could potentially change the sampled state from traced to `NoTrace`. Shouldn't this `if` statement and following overwrite of the `x-request-id` be moved inside the above `if` statement? (or the above if statement changed to do a `return` on `UuidTraceStatus::NoTrace != UuidUtils::isTraceableUuid(x_request_id)`?

If the tracers are ignoring the sampling decision unless starting a new trace, then it wouldn't have an impact - but probably better to not change the sampling decision mid trace in case some tracers do rely upon it.

Reply:
Agree, I'll open a separate issue.